File: dawn_cpp/self_adjusting_list.cpp
#include <iostream>
#include <memory>
#include <optional>

template <typename T>
struct Node {
    std::shared_ptr<Node<T>> prev;
    std::shared_ptr<Node<T>> next;
    T data;
};

// Sleator-Tarjan one-finger model
template <typename T>
class SelfAdjustingList {
private:
    std::shared_ptr<Node<T>> head;
    std::shared_ptr<Node<T>> tail;
    std::shared_ptr<Node<T>> finger;

    std::shared_ptr<Node<T>> traverseFromFinger(const T& target, const std::shared_ptr<Node<T>>& node) {
        if (node->data == target) {
            return node;
        }
        if (node->next == nullptr) {
            finger = tail;
            return nullptr;
        }
        auto found = traverseFromFinger(target, node->next);
        if (found != nullptr) {
            return found;
        }
        finger = node;
        return node;
    }

public:
    SelfAdjustingList() : head(nullptr), tail(nullptr), finger(nullptr) {}

    void insertFront(const T& data) {
        auto newNode = std::make_shared<Node<T>>();
        newNode->data = data;
        if (head == nullptr) {
            head = newNode;
            tail = newNode;
            finger = newNode;
        } else {
            newNode->next = head;
            head->prev = newNode;
            head = newNode;
            finger = newNode;
        }
    }

    std::shared_ptr<Node<T>> find(const T& target) {
        if (finger == nullptr) {
            if (head == nullptr) {
                return nullptr;
            }
            return traverseFromFinger(target, head);
        }
        auto found = traverseFromFinger(target, finger);
        if (found == nullptr) {
            if (head == nullptr) {
                return nullptr;
            }
            return traverseFromFinger(target, head);
        }
        return found;
    }
};



// Constant finger model
template <typename T>
class ConstList {
private:
    std::shared_ptr<Node<T>> head;
    std::shared_ptr<Node<T>> tail;
    std::shared_ptr<Node<T>> finger;

    std::shared_ptr<Node<T>> traverseFromFinger(const T& target, const std::shared_ptr<Node<T>>& node) {
        if (node->data == target) {
            return node;
        }
        if (node->next == nullptr) {
            finger = tail;
            return nullptr;
        }
        auto found = traverseFromFinger(target, node->next);
        if (found != nullptr) {
            return found;
        }
        finger = node;
        return node;
    }

public:
    ConstList() : head(nullptr), tail(nullptr), finger(nullptr) {}

    void insertFront(const T& data) {
        auto newNode = std::make_shared<Node<T>>();
        newNode->data = data;
        if (head == nullptr) {
            head = newNode;
            tail = newNode;
            finger = newNode;
        } else {
            newNode->next = head;
            head->prev = newNode;
            head = newNode;
            finger = newNode;
        }
    }

    std::shared_ptr<Node<T>> find(const T& target) {
        if (finger == nullptr) {
            if (head == nullptr) {
                return nullptr;
            }
            return traverseFromFinger(target, head);
        }
        auto found = traverseFromFinger(target, finger);
        if (found == nullptr) {
            if (head == nullptr) {
                return nullptr;
            }
            return traverseFromFinger(target, head);
        }
        return found;
    }
};


// Order by Next Request strategy
template <typename T>
class ONBRList {
private:
    Node<T>* head;
    Node<T>* tail;

public:
    ONBRList() : head(nullptr), tail(nullptr) {}

    ~ONBRList() {
        Node<T>* curr = head;
        while (curr != nullptr) {
            Node<T>* next = curr->next;
            delete curr;
            curr = next;
        }
    }

    void insert(const T& value) {
        Node<T>* newNode = new Node<T>(value);
        if (head == nullptr) {
            head = newNode;
            tail = newNode;
        } else {
            newNode->next = head;
            head->prev = newNode;
            head = newNode;
        }
    }

    std::optional<T> remove_head() {
        if (head == nullptr) {
            return std::nullopt;
        }
        T value = head->value;
        Node<T>* oldHead = head;
        head = head->next;
        if (head != nullptr) {
            head->prev = nullptr;
        } else {
            tail = nullptr;
        }
        delete oldHead;
        return value;
    }

    void access(const T& value) {
        Node<T>* curr = head;
        Node<T>* prev = nullptr;
        while (curr != nullptr) {
            if (curr->value == value) {
                if (prev == nullptr) {
                    // Value is already at the head
                    return;
                }
                prev->next = curr->next;
                if (curr->next != nullptr) {
                    curr->next->prev = prev;
                } else {
                    tail = prev;
                }
                curr->next = head;
                head->prev = curr;
                head = curr;
                curr->prev = nullptr;
                return;
            }
            prev = curr;
            curr = curr->next;
        }
    }
};

File: dawn_go/self_adjusting_list.go
package main

import "fmt"

// Sleator-Tarjan one-finger model
type Node struct {
	prev   *Node
	next   *Node
	data   interface{}
}

type SelfAdjustingList struct {
	head   *Node
	tail   *Node
	finger *Node
}

func CreateList() *SelfAdjustingList {
	return &SelfAdjustingList{}
}

func (list *SelfAdjustingList) InsertFront(data interface{}) {
	newNode := &Node{data: data}
	if list.head == nil {
		list.head = newNode
		list.tail = newNode
		list.finger = newNode
	} else {
		newNode.next = list.head
		list.head.prev = newNode
		list.head = newNode
		list.finger = newNode
	}
}

func (list *SelfAdjustingList) traverseFromFinger(target interface{}, node *Node) *Node {
	if node.data == target {
		return node
	}
	if node.next == nil {
		list.finger = list.tail
		return nil
	}
	found := list.traverseFromFinger(target, node.next)
	if found != nil {
		return found
	}
	list.finger = node
	return node
}

func (list *SelfAdjustingList) Find(target interface{}) *Node {
	if list.finger == nil {
		if list.head == nil {
			return nil
		}
		return list.traverseFromFinger(target, list.head)
	}
	found := list.traverseFromFinger(target, list.finger)
	if found == nil {
		if list.head == nil {
			return nil
		}
		return list.traverseFromFinger(target, list.head)
	}
	return found
}

// Constant finger model
type ConstList struct {
	head   *Node
	tail   *Node
	finger *Node
}

func CreateList() *ConstList {
	return &ConstList{}
}

func (list *ConstList) InsertFront(data interface{}) {
	newNode := &Node{data: data}
	if list.head == nil {
		list.head = newNode
		list.tail = newNode
		list.finger = newNode
	} else {
		newNode.next = list.head
		list.head.prev = newNode
		list.head = newNode
		list.finger = newNode
	}
}

func (list *ConstList) traverseFromFinger(target interface{}, node *Node) *Node {
	if node.data == target {
		return node
	}
	if node.next == nil {
		list.finger = list.tail
		return nil
	}
	found := list.traverseFromFinger(target, node.next)
	if found != nil {
		return found
	}
	list.finger = node
	return node
}

func (list *ConstList) Find(target interface{}) *Node {
	if list.finger == nil {
		if list.head == nil {
			return nil
		}
		return list.traverseFromFinger(target, list.head)
	}
	found := list.traverseFromFinger(target, list.finger)
	if found == nil {
		if list.head == nil {
			return nil
		}
		return list.traverseFromFinger(target, list.head)
	}
	return found
}


// Order by Next Request strategy
type ONBRList[T any] struct {
	head *Node[T]
	tail *Node[T]
}

func (l *ONBRList[T]) Insert(value T) {
	newNode := &Node[T]{value: value, next: l.head, prev: nil}
	if l.head != nil {
		l.head.prev = newNode
	} else {
		l.tail = newNode
	}
	l.head = newNode
}

func (l *ONBRList[T]) RemoveHead() (T, bool) {
	if l.head == nil {
		var zero T
		return zero, false
	}
	value := l.head.value
	l.head = l.head.next
	if l.head != nil {
		l.head.prev = nil
	} else {
		l.tail = nil
	}
	return value, true
}

func (l *ONBRList[T]) Access(value T) {
	node := l.find(value)
	if node != nil {
		l.moveToFront(node)
	}
}

func (l *ONBRList[T]) find(value T) *Node[T] {
	curr := l.head
	for curr != nil {
		if curr.value == value {
			return curr
		}
		curr = curr.next
	}
	return nil
}

func (l *ONBRList[T]) moveToFront(node *Node[T]) {
	if node == l.head {
		return
	}
	if node.prev != nil {
		node.prev.next = node.next
	} else {
		l.tail = node.next
	}
	if node.next != nil {
		node.next.prev = node.prev
	}
	node.prev = nil
	node.next = l.head
	l.head.prev = node
	l.head = node
}